lib/thrift/parser/models.ex

```diff
@@ -254,6 +254,31 @@ defmodule Thrift.Parser.Models do
 
       %Union{name: name, fields: fields}
     end
+
+    def validator(%Union{}, var_name) do
+      union_var = Macro.var(var_name, nil)
+      quote do
+        set_fields = unquote(union_var)
+        |> Map.delete(:__struct__)
+        |> Enum.reject(fn {_, val} -> is_nil(val) end)
+        case set_fields do
+          [] ->
+            :ok
+
+          [_] ->
+            :ok
+
+          set_fields ->
+            raise Thrift.Union.TooManyFieldsSetException.new(set_fields)
+        end
+      end
+    end
```

Comment:
>  I suspect it would be significantly faster to use the match on the map itself to enforce that only one field is set.

I'd like to see benchmarks that prove that out before any refactoring is explored. In my testing, together, the delete and reject were single-digit microsecond operations.

In addition, the above approach yields the dreaded and unhelpful "NoFunctionClauseError".